### PR TITLE
soc: arm: sam4l: Introduce SoC support

### PR DESCRIPTION
**Pull request by nandojve**
The Atmel SAM4L is a Cortex M4 Ultra Low Power SoC that shares it's peripherals with the sam family. This PR introduces support for the SAM4L MCU and the SAM4L-EK Development Board.

This PR focus on add the basics - DTS, SOC and Board.

What works:

- [ ] GPIO (basic support - soc_gpio_*)
- [X] HWINFO
- [X] USART (Shared Driver)
- [X] SPI (Shared Driver)
- [X] Entropy (Shared Driver)

requires zephyrproject-rtos/hal_atmel/pull/13

### CODE REVIEW

File: boards/arm/sam4l_ek/sam4l_ek.dts
Comment (ioannisg) [inline boards/arm/sam4l_ek/sam4l_ek.dts:12]: We do not need SoC compatibles, here.

File: soc/arm/atmel_sam/sam4l/soc.c
Comment (ioannisg) [inline soc/arm/atmel_sam/sam4l/soc.c:18]: not sure where you need this inclusion

**Review by ioannisg — COMMENTED** [review]
Looks good to me.

File: boards/arm/sam4l_ek/sam4l_ek.dts
Comment (nandojve) [reply boards/arm/sam4l_ek/sam4l_ek.dts:12]: Sorry, it was not clear. Do you want I remove "atmel,sam4l" entry or all the sentence?

File: soc/arm/atmel_sam/sam4l/soc.c
Comment (nandojve) [reply soc/arm/atmel_sam/sam4l/soc.c:18]: I'm trying to do similar things to another soc.c files. My feeling is, If I something need to be dropped here probably there is a bulk update for many soc files that need to be done too.

File: boards/arm/sam4l_ek/sam4l_ek.dts
Comment (ioannisg) [reply boards/arm/sam4l_ek/sam4l_ek.dts:12]: You only need the "board" compatible here, not whatever refers to the SoC or SoC part number

Comment (nandojve) [reply boards/arm/sam4l_ek/sam4l_ek.dts]: done.

File: boards/arm/sam4l_ek/board.cmake
Comment (ioannisg) [inline boards/arm/sam4l_ek/board.cmake:4]: Can you have these properties added in a single line?

File: boards/arm/sam4l_ek/doc/index.rst
Comment (ioannisg) [inline boards/arm/sam4l_ek/doc/index.rst:12]: Can you reword better this phrase:
 ```
The device allows a wide range of options
between functionality and power consumption
```

File: drivers/hwinfo/Kconfig
Comment (ioannisg) [inline drivers/hwinfo/Kconfig:69]: Maybe a different help text as well?

File: boards/arm/sam4l_ek/sam4l_ek_defconfig
Comment (ioannisg) [inline boards/arm/sam4l_ek/sam4l_ek_defconfig:2]: Can you enable ARM_MPU here? Or are you concerned about the amount of image RAM available on this platform?

**Comment by ioannisg** [comment]
@mnkp could you review the GPIO related stuff?

File: boards/arm/sam4l_ek/board.cmake
Comment (nandojve) [reply boards/arm/sam4l_ek/board.cmake:4]: The line will pass 80 char column and I prefer keep this way because of that.

File: boards/arm/sam4l_ek/sam4l_ek_defconfig
Comment (nandojve) [reply boards/arm/sam4l_ek/sam4l_ek_defconfig:2]: Done.

File: boards/arm/sam4l_ek/doc/index.rst
Comment (nandojve) [reply boards/arm/sam4l_ek/doc/index.rst:12]: Done.

File: soc/arm/atmel_sam/sam4l/soc.c
Comment (nandojve) [reply soc/arm/atmel_sam/sam4l/soc.c]: Rem **#include <arch/arm/aarch32/cortex_m/cmsis.h>**

**Comment by nandojve** [comment]
Hi folks, I was wondering if someone had time to do any other check on this PR.
I would like invest time this weekend to fix any issue that anyone can see here.

**Comment by jpsaman** [comment]
@nandojve How far is the support for sam4l s8b now? I am about to start a project with this mcu and am considering using Zephyr for it.

**Comment by nandojve** [comment]
> @nandojve How far is the support for sam4l s8b now? I am about to start a project with this mcu and am considering using Zephyr for it.

Hi @jpsaman, nice to see someone interested on SAM4L. This is first patch to official Zephyr project.
I added basic SoC support that allows you to run at full speed. No power management implement, no roadmap to do it too. I hope community get involved sometime. I made the fundamental that allows enable and disable clocks by peripheral ids. If the full PM is mandatory for you, I do recommend not using SAM4L.

Right now, you can for sure use it with USART, SPI, Entropy, This drivers were on Zephyr for others SAMs and are well validated. I believe that drivers are the only ones that shares IPs with others SAM MCUs. So outside that, I needed re-implement everything because it is a unique IP. However, evetyhing It is at same level from current SAM MCUs including Device Tree and GPIOs. I wrote I2C master driver and an USB (both experimental) on my personal github, you can check here https://github.com/nandojve/zephyr/commits/sam4l_v2.3.0
I've been using JLINK on my boards for development. The board sam4l_wm400_cape (at my personal branch) have partitions to you use SAM-BA, tested on Windows, and it have been working well. So, I believe any SAM4L variation can be used.

Besides I had an 2.2.0 version, my intentions are keep only working on Zephyr mainline.
You are more than welcome to contribute.

NOTE: If you will do a try, don't forget to clone https://github.com/nandojve/hal_atmel/tree/sam4l because you will need symbols to build.

**Comment by jpsaman** [comment]
> Hi @jpsaman, nice to see someone interested on SAM4L. This is first patch to official Zephyr project.
> I added basic SoC support that allows you to run at full speed. No power management implement, no roadmap to do it too. I hope community get involved sometime. I made the fundamental that allows enable and disable clocks by peripheral ids. If the full PM is mandatory for you, I do recommend not using SAM4L.

@nandojve Thanks for giving me an update.

In my project PM is not needed. We do need UART, I2C, GPIO, PWM, NTC and SPI-ETH (ENC424J600) support. How far is this from working? (UART, I2C and GPIO are already ok you say)

**Comment by nandojve** [comment]
> > Hi @jpsaman, nice to see someone interested on SAM4L. This is first patch to official Zephyr project.
> > I added basic SoC support that allows you to run at full speed. No power management implement, no roadmap to do it too. I hope community get involved sometime. I made the fundamental that allows enable and disable clocks by peripheral ids. If the full PM is mandatory for you, I do recommend not using SAM4L.
> 
> @nandojve Thanks for giving me an update.
> 
> In my project PM is not needed. We do need UART, I2C, GPIO, PWM, NTC and SPI-ETH (ENC424J600) support. How far is this from working? (UART, I2C and GPIO are already ok you say)

You need write both an PWM and ADC driver and your on board. ENC424J600 is already on Zephyr tree, so you need define at board dts. I've been using SAM4L to validate RF2XX driver. I recommend 64K SRAM SoC version if you need Zephr native network stack. Using IPv6 with RF2XX 32k is on 99% :D. If you are not using USB, you gain 4K extra memory on a separated address space. So, you have on total 32/64k + 4k.

**Comment by jpsaman** [comment]
> You need write both an PWM and ADC driver and your on board. ENC424J600 is already on Zephyr tree, so you need define at board dts. I've been using SAM4L to validate RF2XX driver. I recommend 64K SRAM SoC version if you need Zephr native network stack. Using IPv6 with RF2XX 32k is on 99% :D. If you are not using USB, you gain 4K extra memory on a separated address space. So, you have on total 32/64k + 4k.

The project uses an sam4l s8b. I see in the zephyr tree a PWM driver named pwm_sam.c, cann't that be reused for sam4l ? The same question for the adc_sam0.c.

**Comment by nandojve** [comment]
> > You need write both an PWM and ADC driver and your on board. ENC424J600 is already on Zephyr tree, so you need define at board dts. I've been using SAM4L to validate RF2XX driver. I recommend 64K SRAM SoC version if you need Zephr native network stack. Using IPv6 with RF2XX 32k is on 99% :D. If you are not using USB, you gain 4K extra memory on a separated address space. So, you have on total 32/64k + 4k.
> 
> The project uses an sam4l s8b. I see in the zephyr tree a PWM driver named pwm_sam.c, cann't that be reused for sam4l ? The same question for the adc_sam0.c.

I not checked if AFEC, from current SAMs, is similar to ADCIFE, from SAM4L, my felling is not equal. I recommend you take look at ASF implementation and compare with Zephyr ones. PWM is a trivial driver. These are less complex drivers and you have a good reference to follow. You worst case scenario is write (adapt) one. To convert an NTC probably low bandwidth is enough and you don't need DMA connected to ADC, basic scenario.

I not see too much risk but you need keep in mind that there is no SAM4L support on Zephyr yet. I hope this can move forward, and when people start to ask, and get involved, things became more relevant and usually get attention. It is normal take time when introducing new SoC but I have hope that this can be merged for 2.4 release.

**Comment by nandojve** [comment]
rebase

File: drivers/hwinfo/hwinfo_sam4l.c
Comment (alexanderwachter) [inline drivers/hwinfo/hwinfo_sam4l.c:25]: a memcpy could be used here. Do you know if we should change the byte order? The datasheet does not give any information on the byte order. What does the asf do here?

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_sam4l.c:25]: From the datasheet:
"Each device has a unique 120 bits serial number readable from address 0x0080020C to0x0080021A."

Comment (nandojve) [reply drivers/hwinfo/hwinfo_sam4l.c]: The docs are not so good.

ASF sample do this:

```
	volatile uint8_t* unique_id_addr = (volatile uint8_t*) UNIQUE_ID_ADDR;

	/* Read the unique id and print it */
	for (i=0; i<15; i++) {
		printf("%02x",*unique_id_addr++);
	}
```
SAM-BA bootloader shows:

```
(sam-ba_2.18) 1 % FLASH::ReadUniqueSerialNumber
buffer address 0x20002a88
Unique Serial Number :
	0x00 : 0x1c
	0x01 : 0xda
	0x02 : 0xa0
	0x03 : 0xc3
	0x04 : 0x62
	0x05 : 0x9f
	0x06 : 0xec
	0x07 : 0xd3
	0x08 : 0x12
	0x09 : 0xd4
	0x0a : 0x9c
	0x0b : 0xbf
	0x0c : 0x6c
	0x0d : 0xdf
	0x0e : 0xec
(sam-ba_2.18) 1 % 
```

I follow same byte order than SAM4/V7 already uses. This way the lower address 0x0080020C will be print as most significant by Zephyr as below:

uart:~$ hwinfo devid 
Length: 15
ID: 0x1cdaa0c3629fecd312d49cbf6cdfec

Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_sam4l.c]: So lets use memcpy and we are good to go.

Comment (nandojve) [reply drivers/hwinfo/hwinfo_sam4l.c]: github is taking time to update files...don't know why. The hash that I sent with memcpy had been showing old version by ~25min. Right now I can see the right version...wierd!

**Review by alexanderwachter — APPROVED** [review]

File: drivers/hwinfo/hwinfo_sam4l.c
Comment (alexanderwachter) [reply drivers/hwinfo/hwinfo_sam4l.c]: Do we need the length < 0. size_t is unsigned anyway.

File: soc/arm/atmel_sam/common/soc_gpio.h
Comment (mnkp) [inline soc/arm/atmel_sam/common/soc_gpio.h:143]: Some small package devices with limited number of pins may have e.g. PIOB and PIOC but not PIOA module. Looking at the SAM SoCs currently supported the above approach should work but it may be safer to test for `PIO_SODR_OFFSET` macro to verify existence of the register. The same holds true for the remaining registers.

**Review by mnkp — CHANGES_REQUESTED** [review]
I propose to remove "drivers: serial: sam: Fix build with old symbols" commit. Atmel header files have traditionally been inconsistent. We can switch to the new style with `_Msk` suffix later when it's better supported. So far the information that the old mask is deprecated (macros without  `_Msk` suffix) is only a comment. The less conditional compilation in the driver code the better.

File: dts/arm/atmel/sam4l.dtsi
Comment (nandojve) [reply dts/arm/atmel/sam4l.dtsi:97]: This is something that I can not change. SAM4L have a very different power management. On the below fragment  is possible see that rule will be different and the shift 6 doesn't apply here.
image

File: dts/arm/atmel/pinctrl_atmel_sam.h
Comment (nandojve) [reply dts/arm/atmel/pinctrl_atmel_sam.h:44]: The SAM4L is a ultra low power design. To achieve that, Atmel optimized SoC on several parts and GPIO is one of that.  This is a unique IP inside SAM's family. The name is `gpio` instead `pio` and class structure is different. This implies that SAM4L will have a dedicated GPIO driver. The macros have a small difference
`atmel,pins = < &pio` vs `atmel,pins = < &gpio`.

File: dts/arm/atmel/sam4l.dtsi
Comment (nandojve) [reply dts/arm/atmel/sam4l.dtsi:55]: Done!

File: soc/arm/atmel_sam/common/soc_gpio.h
Comment (nandojve) [reply soc/arm/atmel_sam/common/soc_gpio.h:143]: Good point! I checked and `PIO_SODR_OFFSET` doesn't exists for SAM4S/E. This means that `PIO_SODR_OFFSET` can't be used. I searched on ASF and noted that ID_GPIO is only defined for SAM4L.
```
#ifdef ID_GPIO
	Gpio *regs;        /** pointer to registers of the GPIO controller */
#else
	Pio *regs;         /** pointer to registers of the PIO controller */
#endif
```

File: boards/arm/sam4l_ek/sam4l_ek.dts
Comment (nandojve) [reply boards/arm/sam4l_ek/sam4l_ek.dts:39]: Done!

File: drivers/serial/usart_sam.c
Comment (nandojve) [reply drivers/serial/usart_sam.c:72]: Done!

File: drivers/serial/usart_sam.c
Comment (nandojve) [reply drivers/serial/usart_sam.c:80]: Removed!

File: drivers/serial/usart_sam.c
Comment (nandojve) [reply drivers/serial/usart_sam.c:82]: Agree!

File: drivers/serial/usart_sam.c
Comment (nandojve) [reply drivers/serial/usart_sam.c:93]: Ok

File: drivers/serial/usart_sam.c
Comment (nandojve) [reply drivers/serial/usart_sam.c]: I think I made all this changes on USART driver when bring up the board. I remember it was not a easy task and I not drop the changes till now.
Removed!

File: dts/arm/atmel/sam4l.dtsi
Comment (mnkp) [reply dts/arm/atmel/sam4l.dtsi:98]: I still think the DTS should reflect the datasheet / header files. If we use different peripheral numbering scheme this will cause maintenance issues. The next person adding a sam4l peripheral to the DTS will use values from the header file. Also, someone struggling to bring up their module and seeing peripheral id not matching the header file will try to chase this trail even if the design was intentional.

It shouldn't take long for anyone to figure out that peripheral ids don't match data sheet on purpose but that's not a user friendly approach.

I would preserve peripheral ids as defined in the header files and modify `soc_pmc_peripheral_enable` function to decode bus_id and peripheral_bit values. Yes, the function code will be a bit more complex.

Comment (nandojve) [reply dts/arm/atmel/sam4l.dtsi]: I made the changes to allow the use of peripheral-id value, defined on the headers. Atmel defined start index (0) at PBAMASK increasing up to PBDMASK, then HSBMASK. For code consistency, I defined CPUMASK as index 5.  So, peripheral-id means at end the clock mask (bus clock) + bit index (peripheral).  Clock mask will be out of order on memory. To solve this, a map translate clock mask group to bus id.

**Review by mnkp — COMMENTED** [review]
Thanks for the update, I have a few more comments.

I propose to remove "dts: watchdog: Add atmel sam TWW watchdog" commit from this PR since it adds DTS files only and we don't have yet the matching watchdog driver. We should not to lose it though. The commit will be useful when implementing the watchdog driver.

File: drivers/serial/usart_sam.c
Comment (nandojve) [reply drivers/serial/usart_sam.c]: Agreed!

File: soc/arm/atmel_sam/common/soc_sam4l_pm.c
Comment (nandojve) [reply soc/arm/atmel_sam/common/soc_sam4l_pm.c:111]: Fixed!

File: soc/arm/atmel_sam/sam4l/soc.c
Comment (nandojve) [reply soc/arm/atmel_sam/sam4l/soc.c]: Sure!

File: soc/arm/atmel_sam/sam4l/soc.c
Comment (nandojve) [reply soc/arm/atmel_sam/sam4l/soc.c:73]: Fixed!

File: soc/arm/atmel_sam/sam4l/soc.c
Comment (nandojve) [reply soc/arm/atmel_sam/sam4l/soc.c]: Fixed!

**Comment by ioannisg** [comment]
@mnkp could you get back to this, and see if we can finalize the review before the freeze?

**Comment by nandojve** [comment]
Rebase and squash "drivers: serial: sam: Add sam4l SoC support" with "soc: arm: atmel: Introduce sam4l SoC" and "dts: arm: atmel: sam4l: Add USART bindings" with "dts: arm: atmel: Introduce sam4l SoC".

**Review by mnkp — APPROVED** [review]

**Comment by mnkp** [comment]
@ioannisg just so it's not overlooked, zephyrproject-rtos/hal_atmel/pull/13 has to be merged first. This PR will still need one more update to fix commit version in west.yml.

File: boards/arm/sam4l_ek/doc/index.rst
Comment (ioannisg) [inline boards/arm/sam4l_ek/doc/index.rst:67]: Can you add here, MPU if it is present?

**Review by ioannisg — CHANGES_REQUESTED** [review]
Just a minor comment in docs.

blocking this until we merge the HAL pr

File: boards/arm/sam4l_ek/doc/index.rst
Comment (nandojve) [reply boards/arm/sam4l_ek/doc/index.rst]: Update table to show on alphabetic order and added missing MPU and HWINFO lines.

**Comment by ioannisg** [comment]
@nandojve can you fix the compliance errors?

**Comment by nandojve** [comment]
> @nandojve can you fix the compliance errors?

It is complex macro for pinctrl. I'm just following what we already have on the main.

**Comment by nandojve** [comment]
Fixed CONFIG_PLL0_* issue.

**Review by ioannisg — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
